File: python/chapter_1/08_zero.py
"""Напишите алгоритм, реализующий условие:
если элемент матрицы MxN равен 0,
то весь столбец и вся строка обнуляются"""
import unittest


def zero(m):
    row = len(m)
    col = len(m[0])
    z = []
    for i in range(row):
        for j in range(col):
            if m[i][j] == 0:
                z.append([i, j])
    for [r, c] in z:
        for i in range(col):
            m[r][i] = 0
        for i in range(row):
            m[i][c] = 0
    return m


class Test(unittest.TestCase):
    data = [([[1, 2, 3, 4, 0],
              [6, 0, 8, 9, 10],
              [11, 12, 13, 14, 15],
              [16, 0, 18, 19, 20],
              [21, 22, 23, 24, 25]],
             [[0, 0, 0, 0, 0],
              [0, 0, 0, 0, 0],
              [11, 0, 13, 14, 0],
              [0, 0, 0, 0, 0],
              [21, 0, 23, 24, 0]])]

    def test_zero(self):
        for [m, result] in self.data:
            self.assertEqual(zero(m), result)


if __name__ == '__main__':
    unittest.main()

File: go/src/chapter_1/08_zero.go
// Напишите алгоритм, реализующий условие:
// если элемент матрицы MxN равен 0,
// то весь столбец и вся строка обнуляются
package chapter_1

func Zero(matrix [][]int) {
	var tmp [][2]int

	m, n := len(matrix), len(matrix[0])
	for row := 0; row < m; row++ {
		for col := 0; col < n; col++ {
			if matrix[row][col] == 0 {
				tmp = append(tmp, [2]int{row, col})
			}
		}
	}

	for _, pos := range tmp {
		for i := 0; i < n; i++ {
			matrix[pos[0]][i] = 0
		}
		for i := 0; i < m; i++ {
			matrix[i][pos[1]] = 0
		}
	}
}
